### PR TITLE
feat: new saga grpc with context

### LINKED ISSUE #454: saga grpc 加上ctx入参
rpc调用时，没有ctx入参会影响到下游rpc服务traceId的生成，最终导致完整链路追踪失效。

### PR DESCRIPTION
**Pull request by Makonike**
close #454 

Due to the involvement of multiple APIs, the preliminary intention is to incorporate the NewSagaGrpcWithContext as an elective invocatory approach. In due course, I shall append illustrative instances and examinations. The subsequent objective encompasses an endeavor to embrace support for opentelemetry.

### CODE REVIEW

**Comment by Makonike** [comment]
@yedf2 
哈喽，由于这里有异步会导致context被cancel，很多测试用例都过不去，所以我希望在这里将上下文拷贝一份使得一些数据不会丢失，因为一些数据已经在Interceptor处理过放到context里了，比如opentelemetry中trace的span、trace啥的，你有什么其他的意见吗？

image

用我瞎写的[Interceptor](https://github.com/asoul-fanclub/jaeger-middleware)试了一下效果，大概像下面这样：

image

用例在分支[feature/context_test](https://github.com/Makonike/dtm/tree/feature/context_test)

File: dtmsvr/trans_process.go [hunk 1]
```diff
@@ -31,14 +32,14 @@ func (t *TransGlobal) process(branches []TransBranch) error {
 	if t.ExtData != "" {
 		dtmimp.MustUnmarshalString(t.ExtData, &t.Ext)
 	}
-
 	if !t.WaitResult {
-		go func() {
+		go func(ctx context.Context) {
+			t.Context = CopyContext(ctx)
```
Comment (yedf2) [inline dtmsvr/trans_process.go:37]: `t.Context = CopyContext(cdx)` should be put before the `go func(ctx context.Context) {`

File: dtmsvr/utils.go [hunk 2]
```diff
@@ -50,3 +54,82 @@ func GetTransGlobal(gid string) *TransGlobal {
...
+	return nil
+}
+
+func (e *timerCtx) String() string {
+	return ""
+}
+
+// CopyContext copy context with value and grpc metadata
+// if raw context is nil, return nil
+func CopyContext(ctx context.Context) context.Context {
```
Comment (yedf2) [inline dtmsvr/utils.go:97]: Is there an elegant way to copy context?

**Comment by yedf2** [comment]
CI failed. In order to run CI on PR, you may make another very simple PR, such as update a comment. After merging it, then you can trigger CI automatically on PR update, not needing approve any more

File: dtmsvr/trans_process.go (hunk 1 above)
Comment (Makonike) [reply dtmsvr/trans_process.go]: OK

File: dtmsvr/utils.go (hunk 2 above)
Comment (Makonike) [reply dtmsvr/utils.go:97]: I will make an attempt to optimize it.

Comment (Makonike) [reply dtmsvr/utils.go:76]: The implementation involving reflection and the unsafe package does indeed raise concerns. In the Go version 1.19.3 used in the CI environment, the structure of timerCtx appears as follows:

image

However, in the Go version 1.20.5 on my local setup, the structure of timerCtx is different, displaying a combination-inheritance composition:

image

In my implementation, I utilize the "unsafe" package to convert a uintptr to a specific timerCtx type. This results in incompatibility between different versions of Go, which can cause issues with the dtm functionality. Do you have any suggestions regarding this?

Comment (Makonike) [reply dtmsvr/utils.go:76]: > Is there an elegant way to copy context?
> 
> I remember that, all values can be overridden in context, so I guess we may pass in another timeout to override the default

No, if the current deadline is already earlier than the new one, no action will be taken.
image

Comment (yedf2) [reply dtmsvr/utils.go:76]: Should not rely on the internal implementation of Context. My suggestion:
```
type asyncCtx struct {
Context
}

func (*emptyCtx) Deadline() (deadline time.Time, ok bool) {
	return
}

func (*emptyCtx) Done() <-chan struct{} {
	return nil
}

func newAysncCtx(parent Context) asyncCtx {
  return asyncCtx{Context: context}
}
```

Comment (yedf2) [reply dtmsvr/utils.go]: why not use anonymous member, so that you do not need to implement Value and Err?

**Pull request merged**